Fix bad sad error when assigning a constant value to an enum member

IMS94 opened the pull request:
## Purpose
When a constant is assigned as the value of an enum member, a bad sad error is printed while internal log complaining of an `IllegalStateException` inside `Desugar`. This PR addresses that issue reported in #27421.

Fixes #27421 

## Approach
### Cause
When `BLangNodeTransformer` handles enum members, it currently treats a constant expression as a simple literal. According to the spec, the constant expression (if available) can be an expression. This cause `Desugar.visit(BLangConstant)` to throw an `IllegalStateException` due to the `BLangConstant#symbol#value#value` being `null`. 

### Fix
- Updated `BLangNodeTransformer` to treat the constant expression (if available) as an expression instead as a simple literal.
- Updated the existing `EnumTest` unit test to cover this scenario as well.

Earlier:
```
    BLangLiteral literal;
    if (member.constExprNode().isPresent()) {
        literal = createSimpleLiteral(member.constExprNode().orElse(null));
```

Now:
```
    if (member.constExprNode().isPresent()) {
        BLangExpression expression = createExpression(member.constExprNode().orElse(null));
```

## Remarks
I have removed the below logic which appeared after obtaining the `BLangLiteral` from either the constant expression or from the member's identifier.
```
        if (literal.originalValue != "" || member.identifier().isMissing()) {
            bLangConstant.setInitialExpression(literal);
        } else {
            bLangConstant.setInitialExpression(createExpression(member.constExprNode().orElse(null)));
        }
```
Since we handle the constant expression (if available) as an expression and handle the identifier as a simple literal (if constant expression is not present), I think we don't need to do an additional check like above. Please check and confirm.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

pubudu91 commented on tests/jballerina-unit-test/src/test/resources/test-src/enums/enums.bal:
```diff
@@ -48,6 +51,9 @@ function testBasicEnumSupport() {
     string q = QUEEN;
     assert(q, "QUEEN");
+
+    string b = BEATLES;
```
Shall we add a test where `BEATLES` is used as a type as well?
e.g., 
```ballerina
BEATLES btl = "The Beatles";
```

IMS94 replied:
Sure. Added.

pubudu91 commented on compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeTransformer.java:
```diff
@@ -3475,19 +3475,15 @@ public BLangConstant transformEnumMember(EnumMemberNode member, Boolean publicQu
-            literal = createSimpleLiteral(member.constExprNode().orElse(null));
+            BLangExpression expression = createExpression(member.constExprNode().orElse(null));
+            bLangConstant.setInitialExpression(expression);
             deepLiteral = createSimpleLiteral(member.constExprNode().orElse(null));
```
This should also be replaced right?

IMS94 replied:
Yes. However, if we do so, how can we rewrite the below logic which occurs in the same method:
```
        if (deepLiteral.originalValue != "") {
            BLangFiniteTypeNode typeNodeAssosiated = (BLangFiniteTypeNode) TreeBuilder.createFiniteTypeNode();
            deepLiteral.originalValue = null;
            typeNodeAssosiated.addValue(deepLiteral);
            bLangConstant.associatedTypeDefinition = createTypeDefinitionWithTypeNode(typeNodeAssosiated);
        } else {
            bLangConstant.associatedTypeDefinition = null;
        }
```

hasithaa commented:
Follow the following guidelines for commit messages. 
https://chris.beams.io/posts/git-commit/

Could you please reword commit messages according to that?

IMS94 commented:
@hasithaa Sorry about that. Squashed both commits to a single commit and updated the commit message according to the guidelines. Could you check if it's ok now?

anupama-pathirage commented:
@IMS94  The commit messages and PR heading is not according to the best practices. Please correct them.

A properly formed Git commit subject line should always be able to complete the following sentence:

If applied, this commit will **your subject line here**

Refer [1] for more details.
 
[1] https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md#submit-your-contribution

IMS94 commented:
@anupama-pathirage I updated the commit message and PR title. Can you check now?
